### ISSUE TITLE
Direction and orientation alerts are needed

### ISSUE THREAD

**Issue by terracoda** [phetsims/friction#116]
As discussed in our design meeting on Tuesday Sept 11, 2018, additional strings for rubbing alerts are ready in the section called, [Direction and Location Alerts](https://docs.google.com/document/d/1v7GM_dNZs86jcRn2X03D2t7HFDb4dQ1h4ayDGXTu63U/edit#heading=h.wl1286o9yftz), in the design document.

@zepumph mentioned he would like to create a sim object type to plug into the friction drag handler. It might have the following options:
- options for defining bounds
- options for managing when to communicate changes in directions - up, left, down, right
- options for what to announce when a bound is reach, e.g., at left edge, at right edge

See design doc for details, and let me know if you have any questions.

**Comment by zepumph** [comment]
In the above commits I added the MovementDescriber to friction. Although it is meant to be a generic type, I kept it in friction and plan to keep it general, perhaps subtyping it if I need more specific model attributes.

The edge alerts work well, but still to do is to
- [ ] make sure that if you keep moving against the wall, the alert will repeat, right now the describer is just based on the position property, so it isn't possible to manage anything with the keybaoard drag listener. 

I also was able to implement a pretty naive strategy for directional alerts, although they are based on the BASE implementation, which involved having diagonal alerts like "up left". I'm not sure we want to support this for friction, but I copied a lot from BASE now, and can hone it in later. 

- [ ] Support diagonal alerts?

Another issue with the directional movement alerts is that, since the describer is just based on the position property, I have been using a threshold to dictate when the object has moved enough to warrant an alert. I would rather just have a single alert when the drag is over.

- [ ] Implement directional alerts for every end drag. Though we may need something like a "direction changed" alert too even if it happens mid drag (left key pressed, right key pressed, left key released, thus no end drag)

**Comment by zepumph** [comment]
I implemented these alerts to now be based on the end drag. Things feel much better. One consideration of note though is that, since I used some implementation from BASE, "left" and "right" are ranges, and not just exactly 0 and 180 degree directions. I think this will be alright, but it will need some testing to see how it converts to this sim.

**Comment by zepumph** [comment]
Still todo beyond the above checkboxes:
- [ ] implement the custom left/right logic as specified in the design doc, and summarized here:
    * if there is air space, always hear all directions
    * if touching, hear "left right" combo once, maybe twice, especially since there are cooling alerts when you break away atoms.

**Comment by terracoda** [comment]
@zepumph, I do not feel the need for for diagonal alerts for Friction, but we do not necessarily need to immediately remove them. 

In BASE diagonal alerts were needed mostly for the free moving balloon after a release. Also, because the draggable area is so much larger in BASE, the diagonal alerts provided a more accurate user experience, including using the mouse where diagonal movement is more natural and more likely to happen.

The desired interaction in Friction is down then left-right, and thankfully, the book does not fly around on its own, so we likely will not need the diagonal alerts for this sim.

**Comment by terracoda** [comment]
@zepumph, this is coming along.
The edges are implemented nicely, I get a sense of how small the rubbing area is which is great.

One issue I noticed is that get incorrect "right" utterances after reaching the top or bottom.

You might have already noticed this, and it might be being caused by the "diagonal logic" left over from BASE.

What happens is that after I reach the top and hear correctly "At top", on all successive Up Arrow key key presses I hear "right" rather than "At top". This is the same for when I reach the bottom. I correctly hear nothing when I reach the Physics book, and then I hear "right", if I press the Down arrow again.

Great work thus far.

**Comment by zepumph** [comment]
See https://github.com/phetsims/friction/issues/123#issuecomment-422990927 for the most up to date links. 

Right now it seems to be working well to only alert one "set" of left/right alerts when the books are touching. If we decide to go with this strategy, then all that is left for this issue is to repeat the edge alerts if you keep trying to move that way.

**Comment by terracoda** [comment]
@zepumph, we will see what participants find. 
We might need at least one "Down" alert in there, even if at the end of that down drag the Chemistry book reaches the Physics book. It often only takes one drag down to get to actually get to the other book, in which case no feedback about going down is spoken. This seemed a little odd to me, especially without the bump sound. 

We can wait for feedback from interviews ;-)

**Comment by terracoda** [comment]
Actually, @zepumph, the incorrect alerts, i.e., "right" on second press at top have already been addressed (just confirmed that).

Based on interview analysis, we definitely need an indication that "down" is important, so I have made adjustments to the strings to help indicate that. This approach is described in https://github.com/phetsims/friction/issues/124 https://github.com/phetsims/friction/issues/125.

The only thing left in this first checkbox in https://github.com/phetsims/friction/issues/116#issuecomment-422600981.

**Comment by zepumph** [comment]
This is the checkbox still to do:

- [ ] make sure that if you keep moving against the wall, the alert will repeat, right now the describer is just based on the position property, so it isn't possible to manage anything with the keybaoard drag listener.

**Comment by zepumph** [comment]
The above checkbox ended up being a large amount of work. In the above commit I added two new classes in a new describer file, and updated the `MovementDescriber` substantially. My goal was to keep the describer generic, and not just usable for friction.

@terracoda please review.

**Comment by zepumph** [comment]
After testing with a screen reader. It is clear that I did a lot of work that is not necessary. I think the main piece that we needed to accomplish was just repeating the alert after every drag, even when we are still against the border. I mistook this to mean that we should be repeating the alert while the drag is happening every second or so. That is what is currently implemented, and it should likely be simplified because the repeated alerts are clipped (at least on NVDA) because there are keys in the downward direction. I will want to discuss this further with @terracoda.

@terracoda when you review, please look at the a11y view and the normal sim with a screen reader. This will help you understanding of how I was repeating the alerts.

**Comment by terracoda** [comment]
@zepumph, left and right alerts are sounding pretty good, but I think you are right, it is better to just fire the alert once, rather than repeat it. 

These kinds of alerts are just to reassure the user that they are successfully moving back and forth and that they have reach an edge and need to change directions. 

Repeating with out a new key press (or drag) might not be re-assuring and could interfere with the more important jiggling alerts. 

In general, I think the repeat alerts at the edges are working except across the top (see checkbox below).

**Issues I found:**
- [ ] I noticed that the "At top" alert fires when moving across the top (using left and right arrows) and when hitting the Top-right edge and the Top-left edge. I don't think we need corner alerts, but ideally,  I would like to hear "At right edge" or "At left edge" when I press the left and right arrow keys or A or D keys, even if also "At top". If easier, we can implement descriptions for the corners. "At top-right" and "At top-left" 

- [ ] I noticed I don't always hear "down", and in some cases I hear "left" or "right" instead of "down". The first couple of downs sound great, though. In these cases, I hear the longer, "Down. Rub fast or slow." After those alerts are exhausted, the down alert seems less predictable.

@zepumph, should I create new issues for the checkbox items?

**Comment by terracoda** [comment]
@zepumph, this version (built version downloaded from PhET Test October 19 12:30 pm CET) sounds the best so far. I only get "Jiggle super fast, super hot"  when there are no more atoms which is perfect and the hint to reset is less prominent and the alerts seem to stop (hit their maximum) on useful content when they stop. 

**We might need:** 
- some tweaking of left-right alerts. I'll start a new issue for that to keep it on the list for us to discuss. 
  - [ ] left-right tweak issue posted
- we might need more explicit (clearer) help text about being on the book to rub - starting an issue for that, too.
 -  [ ] new move book help text posted

**Comment by terracoda** [comment]
Directional and orientation alerts are sounding very good, with potential tweaks to be discussed in #136.

The un-related issue about a change to the move book help text is posted in #137.

@zepumph is it safe to close this issue?

**Comment by terracoda** [comment]
@zepumph, actually, I'll close as the issue is verified. Re-open if you think you need to.
